File: JavascriptAlgorithm/Tree/levelOrder-102.js
var levelOrder1 = function (root) {
  let res = [];
  if (!root) {
    return res;
  }
  let stack = [];
  stack.push(root);
  const traverse = () => {
    while (stack.length > 0) {
      let len = stack.length;
      let tmp = [];
      for (let i = 0; i < len; i++) {
        // 可以用所有指针
        let pop = stack.shift();
        tmp.push(pop.val);
        if (pop.left) {
          stack.push(pop.left);
        }
        if (pop.right) {
          stack.push(pop.right);
        }
      }
      res.push([...tmp]);
    }
  };
  traverse();
  return res;
};

var levelOrder2 = function (root) {
  let res = [];
  if (!root) {
    return res;
  }
  let stack = [];
  stack.push(root);
  const traverse = () => {
    while (stack.length > 0) {
      let len = stack.length;
      let tmp = [];
      for (let i = 0; i < len; i++) {
        // 可以用所有指针
        let pop = stack[i];
        tmp.push(pop.val);
        if (pop.left) {
          stack.push(pop.left);
        }
        if (pop.right) {
          stack.push(pop.right);
        }
      }
      res.push([...tmp]);
      stack.splice(0, len);
    }
  };
  traverse();
  return res;
};
export { levelOrder1, levelOrder2 };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter09/levelOrder_102.java
package algo.tzashinorpu.SecondRound.Chapter09;

import algo.tzashinorpu.SecondRound.TreeNode;

import java.util.LinkedList;
import java.util.List;

public class levelOrder_102 {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        LinkedList<TreeNode> stack = new LinkedList<>();
        stack.offer(root);
        while (stack.size() > 0) {
            int size = stack.size();
            LinkedList<Integer> temp = new LinkedList<>();
            for (int i = 0; i < size; i++) {
                TreeNode pop = stack.pollLast();
                temp.offerLast(pop.val);
                if (pop.left != null) {
                    stack.offerFirst(pop.left);
                }
                if (pop.right != null) {
                    stack.offerFirst(pop.right);
                }
            }
            res.add(temp);
        }
        return res;
    }
}
